docs/architecture.md

```diff
@@ -257,6 +257,67 @@ sequence_batching {
   ID of the sequence in the corresponding batch slot.
 
+#### Implicit State Management
+
+Implicit state management allows a stateful model to store its state inside
+Triton. When using implicit state, the stateful model does not need to store
+the state required for inference inside the model.
+
+Below is a portion of the model configuration that indicates the model
+is using implicit state.
+
+```
+sequence_batching {
+  state [
+    {
+      input_name: "INPUT_STATE"
+      output_name: "OUTPUT_STATE"
+      data_type: TYPE_INT32
+      dims: [ -1 ]
+    }
+  ]
+}
+```
+
+The *state* section in the sequence_batching setting is used to indicate that
+the model is using implicit state. The *input_name* field specifies the name of
+the input tensor that will contain the input state. The *output_name* field
+describes the name of the output tensor produced by the model that contains
+output state. The output state provided by the model in the *i<sup>th</sup>*
+request in the sequence will be used as the input state in the
+*i+1<sup>th</sup>* request. The *dims* field specifies the dimensions of the
+state tensors. When the *dims* field contains variable-sized dimensions, the
+shape of the input state and output state does not have to match.
+
+For debugging purposes, the client can request the output state. In order to
+allow the client to request the output state, the
+[*output* section of the model configuration](./model_configuration.md#inputs-and-outputs)
+must list the output state as one of the model outputs. Note that requesting the
+output state from the client can increase the request latency because of the
+additional tensors that have to be transferred.
+
+Implicit state management requires backend support. Currently, only
+[onnxruntime_backend](https://github.com/triton-inference-server/onnxruntime_backend)
+and [tensorrt_backend](https://github.com/triton-inference-server/tensorrt_backend)
+support implicit state.
+
+##### State Initialization
+
+The starting request in the sequence contains uninitialized data for the input
+state. The model can use the start flag in the request to detect the beginning
+of a new sequence and initialize the model state by providing the initial state
+in the model output. If the *dims* section in the *state* description of the
+model contains variable-sized dimensions, Triton will use *1* for every
+variable-sized dimension for the starting request. For other non-starting
+requests in the sequence, the input state is the output state of the previous
+request in the sequence. For an example ONNX model that uses implicit state to
+accumulate all the inputs in a sequence you can refer to [this CI
+test](../qa/common/gen_qa_implicit_models.py#L130). This model stores the
+partial sum in Triton using implicit state. For state initialization, if the
+request is starting, the model puts the "INPUT" in the "OUTPUT\_STATE" field.
+For non-starting requests, it sets the "OUTPUT\_STATE" and "OUTPUT" to "INPUT"
```

Comment:
IMO no need to talk about OUTPUT here.
> it sets the "OUTPUT\_STATE" to "INPUT" \+  "INPUT\_STATE"

You can elaborate on the model after L313-L315 to what OUTPUT is if you want.